server/auth/registrar.go

```diff
@@ -25,7 +30,372 @@ var (
+	dbAcct, bonds, _, legacyFeePaid := auth.storage.Account(acctID, lockTimeThresh)
+
+	// See if we already have this bond in DB. We still process the request
+	// regardless, but this changes how we proceed.
+	var dbBond *db.Bond
+	for _, bond := range bonds {
+		if bond.AssetID == assetID && bytes.Equal(bond.CoinID, bondCoinID) {
+			dbBond = bond
+			break
+		}
+	}
+	// if dbBond != nil && !dbBond.Pending { /* respond now with dummy confs = reqConfs to save RPCs? */ }
+
+	bondStr := coinIDString(assetID, bondCoinID)
+	bondAssetSym := dex.BipIDSymbol(assetID)
+
+	// Determine if it is pending (found, unconfirmed, fully-confirmed, etc.):
+	// - check confirmations e.g. gettxout
+	// - if tx not found, no error, just confs=-1, respond / break, no DB store (just pre-validate unsigned bond tw)
+	// - store bond info in DB (updating pending flag if already stored)
+	// - if confs >= required, mark bond active and adjust acct tier, no waiter
+	// - if confs < required, start waiter
+	chkAddr, chkAmt, confs, err := auth.checkFee(assetID, bondCoinID)
+	if err != nil {
+		// Not found is expected if this is a pre-broadcast validation request.
+		if !errors.Is(err, asset.CoinNotFoundError) {
+			return msgjson.NewError(msgjson.BondError, "unable to check bond with node: %v", err)
+		}
+
+		// OK. Respond with confs=-1. They may now broadcast, wait for the
+		// required confirms, and postbond again.
+		log.Debugf("Validated prospective bond txn output %s (%s) paying %d for user %v", bondStr, bondAssetSym, amt, acctID)
+		postBondRes.Confs = -1
+		resp, err := msgjson.NewResponse(msg.ID, postBondRes, nil)
+		if err != nil { // shouldn't be possible
+			return msgjson.NewError(msgjson.RPCInternalError, "internal encoding error")
+		}
+		err = conn.Send(resp)
+		if err != nil {
+			log.Warnf("error sending postbond result to link: %v", err)
+		}
+		return nil
+	}
+
+	if int64(chkAmt) != amt {
+		return msgjson.NewError(msgjson.BondError, "node reports incorrect bond amount: %d != %d", chkAmt, amt)
+	}
+
+	if chkAddr != bondAddr {
+		return msgjson.NewError(msgjson.BondError, "node reports incorrect bond address: %v != %v", chkAddr, bondAddr)
+	}
+
+	reqConfs := int64(auth.bondConfs)
+	pending := confs < reqConfs
+
+	// Only store if this is a new bond. If existing, just respond and possibly
+	// start waiting for confirmations and/or flag as active in DB.
+	if dbBond == nil {
+		dbBond = &db.Bond{
+			Version:  postBond.BondVer,
+			AssetID:  assetID,
+			CoinID:   bondCoinID,
+			Script:   postBond.BondScript,
+			Amount:   amt,
+			LockTime: lockTime,
+			Pending:  pending,
+		}
+
+		// The user may be posting their initial bond (new account) or topping up.
+		newAcct := dbAcct == nil
+		if newAcct {
+			log.Infof("Creating new user account %v from %v, posted first bond in %v (%s) / %v",
+				acctID, conn.Addr(), bondStr, bondAssetSym, bondAddr)
+			err = auth.storage.CreateAccountWithBond(acct, dbBond)
```

Comment:
I'll reconsider this, but the interface docs are stale.

Current design is to create the account entry (still tier 0) _as long as the txn is found by the node_, even if unconfirmed.  But it has to be found on the network, thus, the signed and validated bond txn is published.  The idea is that the user can `connect` (tier 0), and then later server can send a `bondconfirmed` notification to the authenticated user once it confirms and `waitBondConf` sees that.

But it must be found by the node.  It won't create accounts based solely on message data anymore like with legacy `'register'`. (See the `auth.checkFee` call above.)

I will reconsider this however.  The client should and does wait until it is fully confirmed before sending the final `postbond` request, but server's node could just be lagging the client's,  so `waitBondConf` is intended to deal with this latency.
I think it should be possible to delay replying with a `PostBondResult` until it's confirmed in the `waitBondConf` goroutine.  I suppose this would make the `bondconfirmed` notification obsolete.

Good suggestion.

Reply:
My main worry was that there could be accounts in the database that never get a bond confirmed. Then they just live on forever. Also there is the possibility of someone creating many transactions that spend the same input but commit to different accounts, and flooding the database with pointless records. Although it would probably be hard to create so many that it causes any issue.

Even if the account is only created when the bond is confirmed, I think there should be a way to clear accounts that have had no active bonds for a while. No reason to keep them around.

Reply:
I do see it your way now.  I did make the changes.

Not sure exactly why I felt the extra complexity of the notifications and the notion of pending bond was worth it, but one challenge deals with the edge case of lost connections after a postbond request is submitted but before a response is received.  I think we can work around that, esp. since if the request is made when client thinks it's fully confirmed, there will be a fairly short period where client is waiting on server to also observe that it's fully confirmed.

> Although it would probably be hard to create so many that it causes any issue.

Next to impossible, I think.  They would be double spends.

> Even if the account is only created when the bond is confirmed, I think there should be a way to clear accounts that have had no active bonds for a while. No reason to keep them around.

That's a somewhat separate issue, but I agree, as long as that account is not deep in the negative tiers on account of major penalties.  We can prune accounts for sure.

Reply:
Circling back on this for clarify.  This change was made in https://github.com/decred/dcrdex/pull/1819/commits/b6bafa324d7c7bf0d6c2e6e9888b2a37e1c7ac1a